src/kudu/consensus/consensus_queue.cc

```diff
@@ -1115,82 +1112,92 @@ int64_t PeerMessageQueue::ComputeNewWatermarkDynamicMode(
   CHECK(local_peer_pb_.has_attrs() && local_peer_pb_.attrs().has_region());
 
   const std::string& leader_region = local_peer_pb_.attrs().region();
-  int total_voters = FindOrDie(voter_distribution, leader_region);
+
+  // Compute the watermarks in leader region. As an example, at the end of this
+  // loop, watermarks_in_leader_region might have entries (3, 7, 5) which
+  // indicates that the leader region has 3 peers that have responded to OpId
+  // indexes 3, 7 and 5 respectively
+  std::vector<int64_t> watermarks_in_leader_region;
+  for (const PeersMap::value_type& peer : peers_map_) {
+    // Only voter members are considered for advancing watermark
+    if (peer.second->peer_pb.member_type() != RaftPeerPB::VOTER) {
+      continue;
+    }
+
+    // Refer to the comment in AdvanceQueueWatermark method for why only
+    // successful last exchanges are considered.
+    if (peer.second->last_exchange_status == PeerStatus::OK) {
+      const string& peer_region = peer.second->peer_pb.attrs().region();
+      if (peer_region == leader_region) {
+        watermarks_in_leader_region.push_back(
+            peer.second->last_received.index());
+      }
+    }
+  }
+
+  int total_voters_from_voter_distribution =
+    FindOrDie(queue_state_.active_config->voter_distribution(), leader_region);
+
+  // Compute number of voters in each region in the active config.
```

Comment:
lines 1141-1163 can be put into active config and reworked once during config update.